code/datums/components/deadchat_control.dm

```diff
@@ -0,0 +1,118 @@
+#define DEMOCRACY_MODE "democracy"
+#define ANARCHY_MODE "anarchy"
+
+/datum/component/deadchat_control
+	dupe_mode = COMPONENT_DUPE_UNIQUE
+	var/timerid
+
+	var/list/datum/callback/inputs = list()
+	var/list/ckey_to_cooldown = list()
+	var/orbiters = list()
+	var/deadchat_mode
+	var/democracy_cooldown
+	var/anarchy_cooldown
+	var/global_control
+
+/datum/component/deadchat_control/Initialize(_deadchat_mode, _inputs, _democracy_cooldown = 12 SECONDS, _anarchy_cooldown = 20 SECONDS, _global_control = FALSE)
+	if(!isatom(parent))
+		return COMPONENT_INCOMPATIBLE
+	RegisterSignal(SSdcs, COMSIG_GLOB_DEAD_SAY, .proc/deadchat_react)
+	RegisterSignal(parent, COMSIG_ATOM_ORBIT_BEGIN, .proc/orbit_begin)
+	RegisterSignal(parent, COMSIG_ATOM_ORBIT_STOP, .proc/orbit_stop)
+	deadchat_mode = _deadchat_mode
+	if(deadchat_mode == DEMOCRACY_MODE)
+		timerid = addtimer(CALLBACK(src, .proc/democracy_loop), democracy_cooldown)
+	inputs = _inputs
+	democracy_cooldown = _democracy_cooldown
+	anarchy_cooldown = _anarchy_cooldown
+	global_control = _global_control
+	if(!global_control)
+		notify_ghosts("[parent] is now deadchat controllable!", source = parent, action = NOTIFY_ORBIT, header="Something Interesting!")
+
+/datum/component/deadchat_control/proc/deadchat_react(datum/source, mob/player, message)
+	if(!global_control)
+		if(!(player in orbiters))
+			return
+	message = lowertext(message)
+	if(!(message in inputs))
+		return 
+	if(deadchat_mode == ANARCHY_MODE)
+		var/cooldown = ckey_to_cooldown[player.ckey]
+		if(cooldown)
+			return SIGNAL_INTERCEPT
+		inputs[message].Invoke()
+		ckey_to_cooldown[player.ckey] = TRUE
+		addtimer(CALLBACK(src, .proc/remove_cooldown, player.ckey), anarchy_cooldown)
+	else if(deadchat_mode == DEMOCRACY_MODE)
+		ckey_to_cooldown[player.ckey] = message
+	return SIGNAL_INTERCEPT
+
+/datum/component/deadchat_control/proc/remove_cooldown(ckey)
+	ckey_to_cooldown.Remove(ckey)
+	
+/datum/component/deadchat_control/proc/democracy_loop()
+	if(QDELETED(parent) || deadchat_mode != DEMOCRACY_MODE)
+		deltimer(timerid)
+		return
+	var/result = count_democracy_votes()
+	if(!isnull(result))
+		inputs[result].Invoke()
+		var/message = "<span class='deadsay italics bold'>[parent] has done action [result]!<br>New vote started. It will end in [democracy_cooldown/10] seconds.</span>"
+		if(global_control)
+			deadchat_broadcast(message)
+		else
+			for(var/M in orbiters)
+				to_chat(M, message)
+	else
+		var/message = "<span class='deadsay italics bold'>No votes were cast this cycle.</span>"
+		if(global_control)
+			deadchat_broadcast(message)
+		else
+			for(var/M in orbiters)
+				to_chat(M, message)
+			
+/datum/component/deadchat_control/proc/count_democracy_votes()
+	var/list/votes = list()
+	for(var/command in inputs)
+		votes["[command]"] = 0
+	var/found_vote = FALSE
```

Comment:
You should be removing the ckeys when you're done so rather than this var you can just do this and check it first thing
```
if(!length(ckey_to_cooldown))
  return
```